core/controller/src/main/scala/whisk/core/controller/actions/PrimitiveActions.scala

```diff
@@ -140,6 +200,387 @@ protected[actions] trait PrimitiveActions {
+   *
+   * Compositions are aborted if the number of action invocations exceeds a limit.
+   * The permitted max is n component invocations plus 2n+1 conductor invocations (where n is the actionSequenceLimit).
+   * The max is chosen to permit a sequence with up to n primitive actions.
+   *
+   * NOTE:
+   * A sequence invocation counts as one invocation irrespective of the number of action invocations in the sequence.
+   * If one component of a composition is also a composition, the caller and callee share the same accounting object.
+   * The counts are shared between callers and callees so the limit applies globally.
+   *
+   * @param components the current count of component actions already invoked
+   * @param conductors the current count of conductor actions already invoked
+   */
+  private case class CompositionAccounting(var components: Int = 0, var conductors: Int = 0)
+
+  /**
+   * A mutable session object to keep track of the execution of one composition.
+   *
+   * NOTE:
+   * The session object is not shared between callers and callees.
+   * A callee has a reference to the session object for the caller.
+   * This permits the callee to return to the caller when done.
+   *
+   * @param activationId the activationId for the composition (ie the activation record for the composition)
+   * @param start the start time for the composition
+   * @param action the conductor action responsible for the execution of the composition
+   * @param cause the cause of the composition (activationId of the enclosing sequence or composition if any)
+   * @param duration the "user" time so far executing the composition (sum of durations for
+   *        all actions invoked so far which is different from the total time spent executing the composition)
+   * @param maxMemory the maximum memory annotation observed so far for the conductor action and components
+   * @param state the json state object to inject in the parameter object of the next conductor invocation
+   * @param accounting the global accounting object used to abort compositions requiring too many action invocations
+   * @param logs a mutable buffer that is appended with new activation ids as the composition unfolds
+   *             (in contrast with sequences, the logs of a hierarchy of compositions is not flattened)
+   * @param caller the session object for the parent composition (caller) if any
+   * @param result a placeholder for returning the result of the composition invocation (blocking invocation only)
+   */
+  private case class Session(activationId: ActivationId,
+                             start: Instant,
+                             action: ExecutableWhiskActionMetaData,
+                             cause: Option[ActivationId],
+                             var duration: Long,
+                             var maxMemory: Int,
+                             var state: Option[JsObject],
```

Comment:
Same comment on mutability.

Reply:
Same answer. I want one Session object for one session. I believe this one to one mapping makes the code easier to follow. Part of what constitutes a session-its state-changes over the course of the execution, so the corresponding fields of the object are mutable.

If I made this class immutable and used object copies instead as suggested, then the distinction between immutable session elements and mutable session elements would disappear from the type and become buried in the code and comments. I would argue this obfuscates intend and hinder readability.